motionplan/plan.go

```diff
@@ -0,0 +1,226 @@
+			newStep[frame] = referenceframe.NewPoseInFrame(pif.Parent(), smuggledGeoPose)
+		}
+		newPath = append(newPath, newStep)
+	}
+	return &rrtPlan{
+		traj:  plan.traj,
+		path:  newPath,
+		nodes: plan.nodes,
+	}, nil
+}
+
+// Trajectory is a slice of maps describing a series of Inputs for a robot to travel to in the course of following a Plan.
+// Each item in this slice maps a Frame's name (found by calling frame.Name()) to the Inputs that Frame should be modified by.
+type Trajectory []map[string][]referenceframe.Input
+
+// GetFrameInputs is a helper function which will extract the waypoints of a single frame from the map output of a trajectory.
+func (traj Trajectory) GetFrameInputs(frameName string) ([][]referenceframe.Input, error) {
+	solution := make([][]referenceframe.Input, 0, len(traj))
+	for _, step := range traj {
+		frameStep, ok := step[frameName]
+		if !ok {
+			return nil, fmt.Errorf("frame named %s not found in trajectory", frameName)
+		}
+		solution = append(solution, frameStep)
+	}
+	return solution, nil
+}
+
+// String returns a human-readable version of the trajectory, suitable for debugging.
+func (traj Trajectory) String() string {
+	var str string
+	for _, step := range traj {
+		str += "\n"
+		for frame, input := range step {
+			if len(input) > 0 {
+				str += fmt.Sprintf("%s: %v\t", frame, input)
+			}
+		}
+	}
+	return str
+}
+
+// EvaluateCost calculates a cost to a trajectory as measured by the given distFunc Metric.
+func (traj Trajectory) EvaluateCost(distFunc ik.SegmentMetric) float64 {
+	var totalCost float64
+	last := map[string][]referenceframe.Input{}
+	for _, step := range traj {
+		for frame, inputs := range step {
+			if len(inputs) > 0 {
+				if lastInputs, ok := last[frame]; ok {
+					cost := distFunc(&ik.Segment{StartConfiguration: lastInputs, EndConfiguration: inputs})
+					totalCost += cost
+				}
+				last[frame] = inputs
+			}
+		}
+	}
+	return totalCost
+}
+
+// Path is a slice of PathSteps describing a series of Poses for a robot to travel to in the course of following a Plan.
+type Path []PathStep
+
+func newRelativePath(solution []node, sf *solverFrame) (Path, error) {
+	path := make(Path, 0, len(solution))
+	for _, step := range solution {
+		inputMap := sf.sliceToMap(step.Q())
+		poseMap := make(map[string]*referenceframe.PoseInFrame)
+		for frame := range inputMap {
+			tf, err := sf.fss.Transform(inputMap, referenceframe.NewPoseInFrame(frame, spatialmath.NewZeroPose()), referenceframe.World)
+			if err != nil {
+				return nil, err
+			}
+			pose, ok := tf.(*referenceframe.PoseInFrame)
+			if !ok {
+				return nil, errors.New("pose not transformable")
+			}
+			poseMap[frame] = pose
+		}
+		path = append(path, poseMap)
+	}
+	return path, nil
+}
+
+func newAbsolutePathFromRelative(path Path) (Path, error) {
+	if len(path) < 2 {
+		return nil, errors.New("need to have at least 2 elements in path")
+	}
+	newPath := make([]PathStep, 0, len(path))
+	newPath = append(newPath, path[0])
```

Comment:
I think we want to explicitly anchor this to `spatialmath.NewZeroPose` rather than implicitly doing so by starting with the first pose.

Reply:
Its not true that this always be Zero in the case of arms for example

Reply:
This ends up being diectly equivalent to `path[0]` anyway and since its a map[string]PoseInFrame means we cant just assign it to the zero pose. Id rather use this shortcut if you dont mind

Reply:
I missed this before but theres nothing in the path spec saying that the first element be zero so this would be misleading